### DISCUSSION TITLE
ConnectionStart, SessionStart safety guarantee.

### DISCUSSION THREAD

**Discussion by Beatle95** [microsoft/msquic#3082] [Q&A] [answered]
Hi, is it safe to call free resources right after ConnectionStart, SessionStart failed? I have something like this in my code:
`
if (QUIC_FAILED(msquic_->ConnectionOpen(registration_, connectionCallback, &connection, &connection.connection_ptr))
        || QUIC_FAILED(msquic_->ConnectionStart(connection.connection_ptr, configuration_, QUIC_ADDRESS_FAMILY_UNSPEC, connection.host.c_str(), connection.port))
    ) {
        msquic_->ConnectionClose(connection.connection_ptr);
    }
...
if (QUIC_FAILED(msquic_->StreamOpen(connection_ptr, QUIC_STREAM_OPEN_FLAG_UNIDIRECTIONAL, streamCallback, &session, &session.session_ptr))
        || QUIC_FAILED(msquic_->StreamStart(session.session_ptr, QUIC_STREAM_START_FLAG_NONE))
    ) {
        msquic_->StreamClose(session.session_ptr);
    }
`
Right now it works just fine, but I don't sure, if it is a good idea.

**Comment by nibanks** [comment, marked as answer]
The only risk is if you try to double free in the "shutdown complete" callbacks; but generally callbacks only happen if "start" was successful, so your current usage pattern seems fine to me.
